core/regexp.c

```diff
@@ -13,48 +13,111 @@ void uwsgi_opt_pcre_jit(char *opt, char *value, void *foobar) {
 #endif
 }
 
-int uwsgi_regexp_build(char *re, pcre ** pattern, pcre_extra ** pattern_extra) {
+int uwsgi_regexp_build(char *re, uwsgi_pcre ** pattern) {
 
+#ifdef UWSGI_PCRE2
+	int errnbr;
+	long unsigned int erroff;
+
+	*pattern = pcre2_compile((const unsigned char *) re, PCRE2_ZERO_TERMINATED, 0, &errnbr, &erroff, NULL);
+#else
 	const char *errstr;
 	int erroff;
 
-	*pattern = pcre_compile((const char *) re, 0, &errstr, &erroff, NULL);
-	if (!*pattern) {
+	*pattern = uwsgi_malloc(sizeof(uwsgi_pcre));
+	(*pattern)->p = pcre_compile((const char *) re, 0, &errstr, &erroff, NULL);
+#endif
+#ifdef UWSGI_PCRE2
+	if (!(*pattern)) {
+		uwsgi_log("pcre error: code %d at offset %d\n", errnbr, erroff);
+#else
+	if (!((*pattern)->p)) {
 		uwsgi_log("pcre error: %s at offset %d\n", errstr, erroff);
+#endif
 		return -1;
 	}
 
+#ifdef UWSGI_PCRE2
+	if (uwsgi.pcre_jit) {
+		errnbr = pcre2_jit_compile(*pattern, PCRE2_JIT_COMPLETE);
+		if (errnbr) {
+			pcre2_code_free(*pattern);
+			uwsgi_log("pcre JIT compile error code %d\n", errnbr);
+			return -1;
+		}
+#else
 	int opt = uwsgi.pcre_jit;
 
-	*pattern_extra = (pcre_extra *) pcre_study((const pcre *) *pattern, opt, &errstr);
-	if (*pattern_extra == NULL && errstr != NULL) {
-		pcre_free(*pattern);
+	(*pattern)->extra = (pcre_extra *) pcre_study((const pcre *) (*pattern)->p, opt, &errstr);
+	if ((*pattern)->extra == NULL && errstr != NULL) {
+		pcre_free((*pattern)->p);
+		free(*pattern);
 		uwsgi_log("pcre (study) error: %s\n", errstr);
 		return -1;
+#endif
 	}
 
 	return 0;
 
 }
 
-int uwsgi_regexp_match(pcre * pattern, pcre_extra * pattern_extra, char *subject, int length) {
-
-	return pcre_exec((const pcre *) pattern, (const pcre_extra *) pattern_extra, subject, length, 0, 0, NULL, 0);
+int uwsgi_regexp_match(uwsgi_pcre *pattern, const char *subject, int length) {
+#ifdef UWSGI_PCRE2
+	return pcre2_match(pattern, (const unsigned char *)subject, length, 0, 0, NULL, NULL);
+#else
+	return pcre_exec((const pcre *) pattern->p, (const pcre_extra *) pattern->extra, subject, length, 0, 0, NULL, 0);
+#endif
 }
 
-int uwsgi_regexp_match_ovec(pcre * pattern, pcre_extra * pattern_extra, char *subject, int length, int *ovec, int n) {
+int uwsgi_regexp_match_ovec(uwsgi_pcre *pattern, const char *subject, int length, int *ovec, int n) {
 
+#ifdef UWSGI_PCRE2
```

Comment:
given the two implemetations share only one line of code i think it would be easier to move the pcre2 version into this block altogether